src/librustc/ty/util.rs

```diff
@@ -1116,6 +1124,8 @@ fn needs_drop_raw<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         ty::TyClosure(def_id, ref substs) => substs.upvar_tys(def_id, tcx).any(needs_drop),
 
+        ty::TyGenerator(..) => true,
```

Comment:
But isn't the destructor only running `drop` on interior types? Maybe you just decided this is a fragile assumption - in that case, please comment.